### PR TITLE
changes to cleanup the volume plugin for recycle

### LINKED ISSUE #26230: Volume plugins cleanup
Filing this for future work.

pkg/volume/plugins.go defines a set of `*VolumePlugin` types that all implement small interfaces like `NewDeleter()`.  This started with `NewMounter()` and `NewUnmounter()` which plugins implement with different structs and are passed around.  It got cargo-culted into a half dozen other such interfaces that do not have the same restrictions, and generally are called like:

```
r := plugin.NewRecycler(name, spec)
r.Recycle()
```

Most of these interfaces could be reduced to:

```
plugin.Recycle(name, spec)
```

Relatively easy one for someone to jump on, but NOT going into the 1.3 milestone.

Additionally, `NewMounter()` and `NewUnmounter()` could be extracted to a `MountableVolumePlugin` type and `VolumeHost` could go through a similar fragmenting.

@jingxu97

### PR DESCRIPTION
**Pull request by kpgriffith**
**What this PR does / why we need it**:
Code cleanup. Changing from creating a new interface from the plugin, that then calls a function to recycle a volume, to adding the function to the plugin itself.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #26230

**Special notes for your reviewer**:
Took same approach from closed PR #28432.

Do you want the approach to be the same for NewDeleter(), NewMounter(), NewUnMounter() and should they be in this same PR or submit different PR's for those?

**Release note**:

```NONE
```

### CODE REVIEW

**Comment by ncdc** [comment]
cc @kubernetes/sig-storage-pr-reviews

**Comment by kpgriffith** [comment]
cc @thockin from issue

**Comment by kpgriffith** [comment]
ping

**Comment by kpgriffith** [comment]
@saad-ali @thockin @jingxu97 Any chance for a review soon?  It's 11 days old now.  Would like to avoid having to close and reopen a new PR.  Thanks.

File: pkg/volume/plugins.go [hunk 1]
```diff
@@ -125,7 +125,7 @@ type RecyclableVolumePlugin interface {
 	// recycler will use the provided recorder to write any events that might be
 	// interesting to user. It's expected that caller will pass these events to
 	// the PV being recycled.
-	NewRecycler(pvName string, spec *Spec, eventRecorder RecycleEventRecorder) (Recycler, error)
+	Recycle(pvName string, spec *Spec, eventRecorder RecycleEventRecorder) error
```
Comment (jingxu97) [inline pkg/volume/plugins.go]: Modify the comments for this change too?

Comment (kpgriffith) [inline pkg/volume/plugins.go]: Done

**Comment by jsafrane** [comment]
I am against this change. All other interfaces in volume plugins have NewMounter() that returns an interface that can do SetUp()/TearDown(), NewAttacher() returns something that can Attach()/Detach(), NewDeleter() returns something that can Delete(), NewProvisioner() returns something that can Provision(), so either change all these or leave NewRecycler() and Recycle() follow the same pattern.

**Comment by kpgriffith** [comment]
@jsafrane Understood.  If you look at the issue, @thockin created, it talks about changing all of them.  In an effort to keep changes small to make it easier to review, I just made a change to Recycle().  If all are good with the implementation, the issue can be left open until all the plugins have changed.  Once this PR is accepted, I plan on making changes to the others.

**Comment by jsafrane** [comment]
Oh, ok.
@k8s-bot ok to test

**Comment by jsafrane** [comment]
lgtm for changes in pkg/controller/volume

**Comment by jsafrane** [comment]
/approve

@jingxu97, @saad-ali, PTAL

**Comment by kpgriffith** [comment]
ping

**Pull request merged**